examples/hosts/app-integration/schema-upgrade/src/containerKillBit.ts

```diff
@@ -57,14 +67,31 @@ export class ContainerKillBit extends DataObject implements IContainerKillBit {
     }
 
     public get markedForDestruction() {
-        return this.crc.read(markedForDestructionKey) as boolean;
+        return this.quorum.get(markedForDestructionKey) as boolean;
     }
 
     public async markForDestruction() {
-        // This should probably use a quorum-type data structure here.
-        // Then, when everyone sees the quorum proposal get approved they can choose to either volunteer
-        // or close themselves
-        await this.crc.write(markedForDestructionKey, true);
+        // Early exit/resolve if already marked.
+        if (this.markedForDestruction) {
+            return;
+        }
+
+        // Note that the marking could come from another client (e.g. two clients try to mark simultaneously).
+        // Watching via the event listener will work regardless of whether our marking or a remote client's
+        // marking was the one that actually wrote the flag.
+        return new Promise<void>((resolve, reject) => {
+            const acceptedListener = (key: string) => {
+                if (key === markedForDestructionKey) {
+                    resolve();
+                    this.quorum.off("accepted", acceptedListener);
+                }
+            };
+            this.quorum.on("accepted", acceptedListener);
+            // Even if quorum.set() becomes a promise, this will remain fire-and-forget since we don't care
+            // whether our marking or a remote client's marking writes the flag (though maybe we'd do retry
+            // logic if a remote client rejects the local client's mark).
+            this.quorum.set(markedForDestructionKey, true);
```

Comment:
So, hypothetically, the proposal could be rejected, and the promise would stay unresolved indefinitely. Do we need a better strategy here involving retries as you mentioned? Perhaps, add a `TODO` here?

Reply:
I think if we add reject functionality to the Quorum, then probably one of two options:
1. The Quorum also gets a `"rejected"` event and here we'll just add a `rejectedListener` (similar to the handling of accept above) which will let us settle the promise.
2. `quorum.set()` does become a promise, but here we only listen for the rejection (acceptance we still take from anyone using the event).

For now I don't intend to add reject unless/until the scenario requires it so it might not ever be an issue.